misc/portless/files/portless.1
.\" Copyright (c) 2006 Martin Kammerhofer
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" @(#)portless.1,v 1.10 2006/08/30 09:36:44 martin Exp
.\"
.\" Note: The date here should be updated whenever a non-trivial
.\" change is made to the manual page.
.Dd August 30, 2006
.Dt PORTLESS 1
.Os
.Sh NAME
.Nm portless
.Nd "quickly browse port descriptions"
.Sh SYNOPSIS
.Nm
.Op Fl dfIiMmp
.Op Fl P Ar pager
.Ar portglob...
.Nm
.Op Fl W | w
.Ar portglob...
.Sh DESCRIPTION
.Nm
lets FreeBSD users quickly browse port descriptions given the port's
name or a (shell type) glob.
.Pp
The options are as follows:
.Bl -tag -width ".Fl P Ar pager"
.It Fl d
Display the
.Pa pkg-descr
file of the port. This is the default.
.It Fl f
Take the
.Ar portglob
argument literal (not as a shell glob).
.It Fl I
Ignore case when matching
.Ar portglob
against the ports tree. This option is only effective if
.Ar portglob
contains at least one unquoted globbing meta-character (
.Sq Li * ,
.Sq Li \&? ,
or
.Sq Li [
).
.It Fl i
Display the
.Pa distinfo
file of the port.
.It Fl M
Display the
.Pa Makefile
file of the port.
.It Fl m
Display the
.Pa pkg-message
file of the port.
.It Fl p
Display the
.Pa pkg-plist
file of the port.
.It Fl P Ar pager
Use the specified
.Ar pager
program to display matching files. This overrides the setting of the
.Ev PAGER
environment variable.
.It Fl w
Rather than browsing files inside the ports tree just echo the names
of directories matching
.Ar portglob
to stdout.
.It Fl W
Like
.Fl w
but echoes just the last two components of the directory (i.e.\&
.Em category/portname
) to stdout.
.El
.Lp
Options
.Fl W
and
.Fl w
are incompatible with all other options.
.Sh ENVIRONMENT
The
.Nm
command uses the following environment variables:
.Bl -tag -width "Ev PORTSDIR" -compact -offset indent
.It Ev PAGER
defaults to
.Dq Pa less\ -e
.It Ev PORTSDIR
defaults to
.Pa /usr/ports
.El
.Sh FILES
.Bl -tag -width ".Pa /usr/ports/*/*/*" -compact
.It Pa /usr/ports/*/*/*
.El
.Sh EXAMPLES
The following are examples of typical usage of the
.Nm
command:
.Pp
.Dl "portless cvsweb\e*"
.Dl "portless lang/\e*doc"
.Dl "portless 'www/*python*'"
.Sh SEE ALSO
.Xr ports_glob 1
from the
.Pa sysutils/portupgrade
port
.Sh AUTHOR
The
.Nm
utility and this manual page were written by
.An Martin Kammerhofer Aq mkamm@gmx.net .
.\" EOF


CHANGES
Updating Information for FreeBSD ports developers

This file is maintained by Erwin Lansing <erwin@FreeBSD.org> and
copyrighted by the FreeBSD Foundation.

This file contains major changes to ports and the ports infrastructure.
Intended audience are ports committers, maintainers and other
developers. User oriented changes should be submitted for inclusion
in the release notes.

All ports committers are allowed to commit to this file.

20060903:
AUTHOR:	sat@FreeBSD.org
The following functional changes were made to bsd.sites.mk:

  * MASTER_SITES variable can now look like MACRO1/subdir1 MACRO2/subdir2, and
    be expanded in a way that you expect it to, that is apply different subdirs
    to different sites. You can safely replace ${M_S_FOO:S/%SUBDIR%/bar/} with
    FOO/bar.

  * If M_S_SUBDIR is unset, a macro in M_S won't set it anymore, but rather
    substitute %SUBDIR% all by itself. Now it's possible to use macros with
    different default subdirs and they will expand in a proper way.

20060805:
AUTHOR:	sat@FreeBSD.org
The following functional changes were made to bsd.sites.mk:

  * MASTER_SITES variable can now contain special macros which are expanded into
    traditional URL[:group] syntax and cause some convenient effects, e.g. set a
    default M_S_SUBDIR.

  * A macro is any space-delimited word in MASTER_SITES without a forward slash.

  * If MASTER_SITE_<macro> is defined, then the macro is expanded to its value.

  * Abbreviated macros are provided for extremely popular locations, e.g. CPAN
    and SF. To avoid further obfuscation their unabbreviated equivalents should
    be avoided.

  * Popular macros can trigger M_S_SUBDIR to default to a predefined value, e.g.
    ${PORTNAME:L} for SF and ${PORTNAME:C/-.*//} for CPAN.

  * Abbreviations and M_S_SUBDIR defaults are defined by MASTER_SITES_ABBREVS
    and MASTER_SITES_SUBDIRS, both set in bsd.sites.mk.

  * If a macro belongs to a group (e.g. SF:source1), all the sites it expands
    into also belong to the same group.

  * Macros and traditional URL's can be mixed safely, their order stays intact.

20060804:
AUTHOR:	gabor@FreeBSD.org
The following functional changes were made to bsd.*.mk and to 
Tools/scripts/security-check.awk:

  * Add DESTDIR support to let one install ports into a jail from outside.
    A package is installed under ${DESTDIR}${PREFIX} now and registered under
    ${DESTDIR}/var/db/pkg.

  * TARGETDIR variable has been added for referencing ${DESTDIR}${PREFIX}.

  * LOCALBASE, LINUXBASE and X11BASE variables point to the final destination
    now, LOCALBASE_REL, LINUXBASE_REL and X11BASE_REL can be used for the
    old behavior.  The behavior has not been changed in PLIST_SUB.
    Note, that the three variables are still overrideable with a hack, so this
    change doesn't affect POLA, but necessary for DESTDIR support.

  * OSVERSION is now determined from the userland, not from the kernel.
    Cross-compiling between releases is still not supported, but this
    behavior is more correct for jails with differing userland from
    the kernel version.

  * Add some sanity check for PREFIX and DESTDIR.  This can be skipped by
    defining IGNORE_PATH_CHECKS, but very discouraged.

  * Deinstalling from DESTDIR, determining if a port is already installed in
    DESTDIR, etc. are implemented by chrooting those specific commands.

  * Text changes to reflect that we are installing to DESTDIR, removing from
    DESTDIR, etc.

Note, that this change implements only the infrastructure support.  Ports should
also be implemented to respect DESTDIR.

20060717:
AUTHOR: portmgr@FreeBSD.org
The following functional changes were made to bsd.port.mk:

  * Introduces IA32_BINARY_PORT for certain cases where a given port fetches
    and installs compiled i386 binaries.

  * Adds some infrastructure for support of 32-bit i386 apps on amd64.

  * Again fix USE_LDCONFIG logic to make it work with non-default PREFIX to
    fix bug introduced in previous revision.

20060705:
AUTHOR: portmgr@FreeBSD.org
The following functional changes were made to bsd.port.mk:

  * Change all bogus uses of BROKEN to IGNORE.  (Note: the BROKEN_WITH_*
    forms are retained for compatibility but deprecated.)  Internally,
    these variables all had already set IGNORE, not BROKEN, so they
    were really misnomers.

  * Fix USE_LDCONFIG with non-default PREFIX.

  * Fix DESKTOP_ENTRIES processing on 4.x.

  * Add 'make missing' to show missing dependencies.

  * Add bsd.scons.mk and bsd.wx.mk.

20060616:
AUTHOR: portmgr@FreeBSD.org
  The patch in ports/95841 to get rid of remnants of FORCE_PKG_REGISTER
  was backed out because it broke kdelibs3*.

20060616:
AUTHOR: portmgr@FreeBSD.org
The following functional changes were made to bsd.port.mk:

  * Update default MySQL version from 4.1 to 5.0 [1]

  * Update default PHP version from 4 to 5 [2]

  * Update default LDAP version from 2.2 to 2.3 [3]

  * Add support for LDAP version 2.4 [4]

  * Update default linux_base from '8' (Redhat 8) to 'fc4' (Fedora Core 4) [5]

  * Switch default Linux X implementation from XFree86 to xorg to track the
    above [5]

  * Deprecate INSTALLS_SHLIB and replace it by USE_LDCONFIG.  If set to "yes",
    the old behavior is preserved.  Otherwise, it can be set to a list of
    directories to be added to ${PREFIX}/libdata/ldconfig/${UNIQUENAME}.
    Note that this directory is used by ldconfig startup script, it is meant
    to replace ldconfig scripts installed by some ports as (sometimes
    000.${UNQUENAME}.sh) [6]

  * Adds USE_LDCONFIG32 which is like USE_LDCONFIG but the target file is
    ${PREFIX}/libdata/ldconfig32/${UNIQUENAME} instead.  (Note: this should
    only be used on 64-bit architectures) [6]

  * Set the default LANG for tr to be C to avoid some build problems [7]

  * Include bsd.java.mk in both pre and post sections [8]

  * Get rid of remnants of FORCE_PKG_REGISTER [9]

  * Always print options with make showconfig [10]


ports-mgmt/portless/files/portless.sh
#! @BASH@ --
# (X)Emacs: -*- mode: Shell-Script; coding: iso8859-1; -*-
# @(#)portless.sh,v 1.12 2006/08/30 09:21:22 martin Exp
# Show "pkg-descr" file of matching port(s).
#
# Copyright (c) 2006 Martin Kammerhofer <mkamm@gmx.net>
# All rights reserved.
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.

Script=`basename $0` # name of this script

# set defaults
for opt in d f I i M m P p W w; do
    eval opt_$opt=""
done
PORTSDIR=${PORTSDIR:-/usr/ports}
PAGER=${PAGER:-less -e}
PKGDESCR="pkg-descr"
filelist=""
rc=0

# print usage message to stderr and exit
usage()
{
    while [ $# != 0 ]; do
	echo >&2 "$Script: $1"
	shift
    done
    echo >&2 "usage: $Script [-dfIiMmp | -Ww] [-P pager] 'portglob'..."
    exit 64	# EX_USAGE
}

fatal() { echo >&2 "$Script: $*"; exit 70; }	# EX_SOFTWARE

# Add file $2 corresponding to option $1 to $filelist.
# Do not add any file twice.
addopt()
{
    eval "[ -n \"\$opt_$1\" ] && return; opt_$1='$1' || fatal 'addopt $*'"
    filelist="$filelist $2"
}

# process options
while getopts "D:dfIiMmpP:Wwx" option
  do
  case "$option" in
      (D) PORTSDIR="$OPTARG";;		# undocumented
      (d) addopt d "$PKGDESCR";;
      (f) opt_f="f";;
      (I) shopt -s nocaseglob ||
      usage "option -I needs bash!"
      opt_I="I";;
      (i) addopt i "distinfo";;
      (M) addopt M "Makefile";;
      (m) addopt m "pkg-message";;
      (p) addopt p "pkg-plist";;
      (P) PAGER="$OPTARG"; opt_P="P";;
      (W) opt_W="W";;
      (w) opt_w="w";;
      (x) set -x;;			# undocumented
      (*) usage;;
  esac
done
shift $(($OPTIND - 1))
[ -d "$PORTSDIR" ] || fatal "No such directory '$PORTSDIR'"
if [ -n "$opt_W" -o -n "$opt_w" ]; then
    [ -n "$filelist" -o -n "$opt_P" -o -n "$opt_W" -a -n "$opt_w" ] &&
	usage "options -W and -w are not compatible with any other option!"
    PAGER="/bin/ls -1d"			# just echo directory name(s)
    filelist="."
elif [ -n "$opt_f" -a -n "$opt_I" ]; then
    usage "only one of options -f and -I is useful!"
else
    [ -n "$filelist" ] || filelist="$PKGDESCR"
fi

# there must be at least one argument
if [ $# = 0 ]; then
    usage
fi

# View $PORTSDIR$1$2/{$filelist} with $PAGER.
# Always glob $1 but do not glob $2 when option -f was given.
# (Slightly complicated because /bin/sh may not understand
# csh-like {file1,file2} glob patterns.)
page() {
    [ $# = 2 ] || fatal "page $*"
    dirglob=$1
    portglob=$2
    shift 2
    for f in $filelist; do
	if [ -n "$opt_f" ]; then
	    set -- "$@" "$PORTSDIR"$dirglob"$portglob/$f"
	else
	    set -- "$@" "$PORTSDIR"$dirglob$portglob/"$f"
	fi
    done
    if [ -n "$opt_W" -o -n "$opt_w" ]; then
	lstrip=""
	[ -n "$opt_W" ] && lstrip=-e\ 's;^.*/\([^/]*/[^/]*\)$;\1;'
	{
	    $PAGER "$@" | grep -Ev '/distfiles/|/packages/' |
	    sed -e 's;/.$;;' $lstrip
	} || { rc=$?; return $rc; }		# remember last error code
    else
	$PAGER "$@" || { rc=$?; return $rc; }	# remember last error code
    fi
}

# main
for p in "$@"; do
    expr "$p" : ".*[$IFS]" >/dev/null &&
	usage "portglob '$p' contains whitespace!"
    case "$p" in
	(*/*/*)
	usage "portglob '$p' contains more than one slash!";;

	(*/*)
	page "/" "$p";;

	(*)
	page "/[a-z]*/" "$p";;
    esac
done

exit $rc

#EOF#
